File: test/extensions/language/expressions/python/BitmaskTypeTest.py
import Expressions


class BitmaskTypeTest(Expressions.TestCase):
    def testBitSizeOfNoColor(self):
        bitmaskTypeExpression = self.api.BitmaskTypeExpression()
        bitmaskTypeExpression.colors = self.api.Colors()
        bitmaskTypeExpression.has_not_color_red = True

        self.assertEqual(9, bitmaskTypeExpression.bitsizeof())

    def testBitSizeOfRed(self):
        bitmaskTypeExpression = self.api.BitmaskTypeExpression()
        bitmaskTypeExpression.colors = self.api.Colors.Values.RED
        bitmaskTypeExpression.has_color_red = True

        self.assertEqual(9, bitmaskTypeExpression.bitsizeof())

    def testBitSizeOfGreen(self):
        bitmaskTypeExpression = self.api.BitmaskTypeExpression()
        bitmaskTypeExpression.colors = self.api.Colors.Values.GREEN
        bitmaskTypeExpression.has_color_green = True
        bitmaskTypeExpression.has_not_color_red = True
        bitmaskTypeExpression.has_other_color_than_red = True

        self.assertEqual(11, bitmaskTypeExpression.bitsizeof())

    def testBitSizeOfBlue(self):
        bitmaskTypeExpression = self.api.BitmaskTypeExpression()
        bitmaskTypeExpression.colors = self.api.Colors.Values.BLUE
        bitmaskTypeExpression.has_color_blue = True
        bitmaskTypeExpression.has_not_color_red = True
        bitmaskTypeExpression.has_other_color_than_red = True

        self.assertEqual(11, bitmaskTypeExpression.bitsizeof())

    def testBitSizeOfBlueGreen(self):
        bitmaskTypeExpression = self.api.BitmaskTypeExpression()
        bitmaskTypeExpression.colors = self.api.Colors.Values.BLUE | self.api.Colors.Values.GREEN
        bitmaskTypeExpression.has_color_blue = True
        bitmaskTypeExpression.has_color_green = True
        bitmaskTypeExpression.has_not_color_red = True
        bitmaskTypeExpression.has_other_color_than_red = True

        self.assertEqual(12, bitmaskTypeExpression.bitsizeof())

    def testBitSizeOfAllColors(self):
        bitmaskTypeExpression = self.api.BitmaskTypeExpression()
        bitmaskTypeExpression.colors = (
            self.api.Colors.Values.RED | self.api.Colors.Values.BLUE | self.api.Colors.Values.GREEN
        )
        bitmaskTypeExpression.has_color_red = True
        bitmaskTypeExpression.has_color_blue = True
        bitmaskTypeExpression.has_color_green = True
        bitmaskTypeExpression.has_all_colors = True
        bitmaskTypeExpression.has_other_color_than_red = True

        self.assertEqual(13, bitmaskTypeExpression.bitsizeof())

File: test/extensions/language/expressions/cpp/BitmaskTypeTest.cpp
#include "expressions/bitmask_type/BitmaskTypeExpression.h"
#include "gtest/gtest.h"

namespace expressions
{
namespace bitmask_type
{

TEST(BitmaskTypeTest, bitSizeOfNoColor)
{
    BitmaskTypeExpression bitmaskTypeExpression;
    bitmaskTypeExpression.setColors(Colors());
    bitmaskTypeExpression.setHasNotColorRed(true);

    ASSERT_EQ(9, bitmaskTypeExpression.bitSizeOf());
}

TEST(BitmaskTypeTest, bitSizeOfRed)
{
    BitmaskTypeExpression bitmaskTypeExpression;
    bitmaskTypeExpression.setColors(Colors::Values::RED);
    bitmaskTypeExpression.setHasColorRed(true);

    ASSERT_EQ(9, bitmaskTypeExpression.bitSizeOf());
}

TEST(BitmaskTypeTest, bitSizeOfGreen)
{
    BitmaskTypeExpression bitmaskTypeExpression;
    bitmaskTypeExpression.setColors(Colors::Values::GREEN);
    bitmaskTypeExpression.setHasColorGreen(true);
    bitmaskTypeExpression.setHasNotColorRed(true);
    bitmaskTypeExpression.setHasOtherColorThanRed(true);

    ASSERT_EQ(11, bitmaskTypeExpression.bitSizeOf());
}

TEST(BitmaskTypeTest, bitSizeOfBlue)
{
    BitmaskTypeExpression bitmaskTypeExpression;
    bitmaskTypeExpression.setColors(Colors::Values::BLUE);
    bitmaskTypeExpression.setHasColorBlue(true);
    bitmaskTypeExpression.setHasNotColorRed(true);
    bitmaskTypeExpression.setHasOtherColorThanRed(true);

    ASSERT_EQ(11, bitmaskTypeExpression.bitSizeOf());
}

TEST(BitmaskTypeTest, bitSizeOfBlueGreen)
{
    BitmaskTypeExpression bitmaskTypeExpression;
    bitmaskTypeExpression.setColors(Colors::Values::BLUE | Colors::Values::GREEN);
    bitmaskTypeExpression.setHasColorGreen(true);
    bitmaskTypeExpression.setHasColorBlue(true);
    bitmaskTypeExpression.setHasNotColorRed(true);
    bitmaskTypeExpression.setHasOtherColorThanRed(true);

    ASSERT_EQ(12, bitmaskTypeExpression.bitSizeOf());
}

TEST(BitmaskTypeTest, bitSizeOfAllColors)
{
    BitmaskTypeExpression bitmaskTypeExpression;
    bitmaskTypeExpression.setColors(Colors::Values::RED | Colors::Values::GREEN | Colors::Values::BLUE);
    bitmaskTypeExpression.setHasColorRed(true);
    bitmaskTypeExpression.setHasColorGreen(true);
    bitmaskTypeExpression.setHasColorBlue(true);
    bitmaskTypeExpression.setHasAllColors(true);
    bitmaskTypeExpression.setHasOtherColorThanRed(true);

    ASSERT_EQ(13, bitmaskTypeExpression.bitSizeOf());
}

} // namespace bitmask_type
} // namespace expressions
